### DISCUSSION TITLE
How to set Compose key?

### DISCUSSION THREAD

**Discussion by TheWhiteFangedOne** [hyprwm/Hyprland#1224] [Q&A]
I am unsure how to set a Compose key like I could in Sway, using xkb_options. Using setxkbmap simply returned an error about running xkb against Xwayland.

**Comment by vaxerski** [comment]
isn't compose just a mod? you can see that in `wev`

**Reply by xlassmgreat** [reply]
What wev? Please explain.

**Reply by vaxerski** [reply]
https://git.sr.ht/~sircmpwn/wev

**Reply by xlassmgreat** [reply]
Of course, a wayland version of xev.  I don't know why I didn't really but, okay. What exactly am I supposed to do with wev? What am I looking for? So you mean the compose key is a mod key, but what if I'd need to change that to something else?

**Reply by vaxerski** [reply]
when you press a mod, you'll get a mod event with its name:
```
[14:     wl_keyboard] modifiers: serial: 0; group: 0
                      depressed: 00000040: Mod4 
                      latched: 00000000
                      locked: 00000000
```

if you want to remap it to something, you have the kb_<something> variables in the hyprland config to input your xkb layouts options and stuff. Hyprland should not have to deal with remapping and stuff.

**Reply by xlassmgreat** [reply]
I don't understand where you're going with this. I'm not trying to remap anything to anything. I'm trying to set the compose key, how would I approach it? Normally, I'd expect setting `kb_options = "compose:rctrl"` would set the compose key to the right control key. How does this work?

Edit: It's been more than a year, and I ought to have mentioned it sooner. But I'll say it now: my mistake was using double quotes. Silly as that.

**Reply by prskr** [reply]
at least for me it works actually more or less exactly like you're describing it:

```
input {
    kb_layout = us
    kb_variant =
    kb_model =
    kb_options = grp:alt_shift_toggle, compose:menu
}
```

works perfectly so use the menu key as compose key

**Reply by Schievel1** [reply]
> at least for me it works actually more or less exactly like you're describing it:
> 
> ```
> input {
>     kb_layout = us
>     kb_variant =
>     kb_model =
>     kb_options = grp:alt_shift_toggle, compose:menu
> }
> ```
> 
> works perfectly so use the menu key as compose key

Thanks. This works for ralt, but not for lalt or lsuper for some reason. I would like to bind it to 'page up'. What would be the right keycode for that? It's not 'pgup' it seems.

**Reply by Kenha09** [reply]
finally I found the solution. I was also searching for a way to use the Compose keys as my KDE plasma settings didn't look much as the tutorials I've seen. And I'm using Arch + Hyprland + wayland + .config file. The 'kb_options = compose:rctrl' fixed my life. Now I can add the spanish symbols without any problem ñ, á, etc! very happy, thanks a lot!

**Comment by Kantariia** [comment]
Sorry, bro. I am from Ukraine. Now i at the war. I can't open my code.
So i am So sorry.

вс, 8 янв. 2023 г., 07:15 xlassmgreat ***@***.***>:

**Comment by gurgelgubbe** [comment]
This works wonders for me :)

```
set {
  $keyboard_layout  se,se
  $keyboard_model   pc105
  $keyboard_options grp:alt_space_toggle,compose:rctrl
  $keyboard_variant us,
}

input * {
  xkb_layout  $keyboard_layout
  xkb_variant $keyboard_variant
  xkb_options $keyboard_options
  xkb_model   $keyboard_model
}

```

**Comment by wideaperture** [comment]
I anyone's still looking for an answer as to how to set the compose key (not the compose menu),  I happened across it [here](https://autumns.page/posts/2022/12/hyprland-compose/). Setting the compose key to the right alt key can be done like so in `hyprland.conf`:

```
input {
    # ...
    kb_options = compose:ralt
    # ...
}
```

**Comment by Schievel1** [comment]
Still, how do you actually get those keycodes? wev gives this when pressing left alt key: 
```
[14:     wl_keyboard] key: serial: 8137; time: 4871790; key: 64; state: 1 (pressed)
                      sym: Alt_L        (65513), utf8: ''
[14:     wl_keyboard] modifiers: serial: 0; group: 0
                      depressed: 00000008: Mod1
                      latched: 00000000
                      locked: 00000000
[14:     wl_keyboard] key: serial: 8139; time: 4872257; key: 64; state: 0 (released)
                      sym: Alt_L        (65513), utf8: ''
[14:     wl_keyboard] modifiers: serial: 0; group: 0
                      depressed: 00000000
                      latched: 00000000
                      locked: 00000000
```
In what way does that help in guessing what the right thing is to put after `compose:`?

For right alt its `ralt`, but for left alt it is not `lalt` apparently. `ralt` works for me, so does `caps`. But not `pageup`

Edit: To answer that question: 
The allowed configurations are found in a file:
```
guru/app-editors/emacs-lsp-booster on  dev [$] ➜ grep "compose:" /usr/share/X11/xkb/rules/base.lst   took 11m48s
  compose:ralt         Right Alt
  compose:lwin         Left Win
  compose:lwin-altgr   3rd level of Left Win
  compose:rwin         Right Win
  compose:rwin-altgr   3rd level of Right Win
  compose:menu         Menu
  compose:menu-altgr   3rd level of Menu
  compose:lctrl        Left Ctrl
  compose:lctrl-altgr  3rd level of Left Ctrl
  compose:rctrl        Right Ctrl
  compose:rctrl-altgr  3rd level of Right Ctrl
  compose:caps         Caps Lock
  compose:caps-altgr   3rd level of Caps Lock
  compose:102          The "&lt; &gt;" key
  compose:102-altgr    3rd level of the "&lt; &gt;" key
  compose:paus         Pause
  compose:ins          Insert
  compose:prsc         PrtSc
  compose:sclk         Scroll Lock
```

**Comment by opennomad** [comment]
One thing that caught me is that I tried to add the `kb_options` twice. Once for swapping caps and escape and again for mapping the compose. The right way to do it is to put them on the same line and separate them with a comma, as follows:
```
kb_options = caps:swapescape,compose:rctrl
```
That makes it so I can press the right control to activate compose and type things like ä (compose + " + a) and ⁰ (compose + ^ + 0) and ñ (compose + ~ + n) etc pressing one key at a time.
